circuits/src/types/fee.rs

```diff
@@ -543,3 +556,173 @@ impl<N: MpcNetwork + Send, S: SharedValueSource<Scalar>> CommitVerifier
+}
+
+impl Add<FeeSecretShare> for FeeSecretShare {
+    type Output = Fee;
+
+    fn add(self, rhs: FeeSecretShare) -> Self::Output {
+        let settle_key = scalar_to_biguint(&(self.settle_key + rhs.settle_key));
+        let gas_addr = scalar_to_biguint(&(self.gas_addr + rhs.gas_addr));
+        let gas_token_amount = scalar_to_u64(&(self.gas_token_amount + rhs.gas_token_amount));
+        let percentage_fee = FixedPoint::from(self.percentage_fee + rhs.percentage_fee);
+
+        Fee {
+            settle_key,
+            gas_addr,
+            gas_token_amount,
+            percentage_fee,
+        }
+    }
+}
+
+impl FeeSecretShare {
+    /// Apply a blinder to the secret shares
+    pub fn blind(&mut self, blinder: Scalar) {
+        self.settle_key += blinder;
+        self.gas_addr += blinder;
+        self.gas_token_amount += blinder;
+        self.percentage_fee += blinder;
+    }
+
+    /// Remove a blinder from the secret shares
+    pub fn unblind(&mut self, blinder: Scalar) {
+        self.settle_key -= blinder;
+        self.gas_addr -= blinder;
+        self.gas_token_amount -= blinder;
+        self.percentage_fee -= blinder;
+    }
+}
+
+/// Represents a fee secret share that has been allocated in a constraint system
+#[derive(Clone, Copy, Debug)]
+pub struct FeeSecretShareVar {
+    /// The public settle key of the cluster collecting fees
+    pub settle_key: Variable,
+    /// The mint (ERC-20 Address) of the token used to pay gas
+    pub gas_addr: Variable,
+    /// The amount of the mint token to use for gas
+    pub gas_token_amount: Variable,
+    /// The percentage fee that the cluster may take upon match
+    /// For now this is encoded as a u64, which represents a
+    /// fixed point rational under the hood
+    pub percentage_fee: Variable,
+}
+
+impl Add<FeeSecretShareVar> for FeeSecretShareVar {
+    type Output = FeeVar<LinearCombination>;
+
+    fn add(self, rhs: FeeSecretShareVar) -> Self::Output {
+        FeeVar {
+            settle_key: self.settle_key + rhs.settle_key,
+            gas_addr: self.gas_addr + rhs.gas_addr,
+            gas_token_amount: self.gas_token_amount + rhs.gas_token_amount,
+            percentage_fee: self.percentage_fee + rhs.percentage_fee,
+        }
+    }
+}
+
+impl FeeSecretShareVar {
+    /// Apply a blinder to the secret shares
+    pub fn blind(&mut self, blinder: Variable) {
+        self.settle_key += blinder;
+        self.gas_addr += blinder;
+        self.gas_token_amount += blinder;
+        self.percentage_fee += blinder;
+    }
+
+    /// Remove a blinder from the secret shares
+    pub fn unblind(&mut self, blinder: Variable) {
+        self.settle_key -= blinder;
+        self.gas_addr -= blinder;
+        self.gas_token_amount -= blinder;
+        self.percentage_fee -= blinder;
+    }
+}
+
+/// Represents a commitment to a fee secret share that has been allocated in a constraint system
+#[derive(Clone, Copy, Debug)]
+pub struct FeeSecretShareCommitment {
```

Comment:
just thinking out loud... don't know if this is a feature or a nuisance of Rust but we have these `X`, `XSecretShare`, `XSecretShareVar`, `XSecretShareCommitment` structs that all have the same field names (different types ofc), which feels annoying to redefine all of them... different impls too i know but still.

maybe, down the line, we can make a macro that derives these structs from a base `X` struct? It seems like `XSecretShare`, `XSecretShareVar`, `XSecretShareCommitment` all have fields of the same type (`Scalar`, `Variable`, `CompressedRistretto`, respectively), which would make this easier.

but definitely something to do down the line

Reply:
Yeah this would be nice, and definitely something to consider, I'll paint some color into why the `circuits` crate is written this way.

Everything in the circuits crate is pretty verbose; we only really define traits to define interfaces for common data manipulation patterns, (i.e. commit, prove, verify). It's written this way to be maximally clear, auditable, and make bugs stick out more.

I think introducing macros has a convenience advantage at a high readability cost, and for the circuits crate so far I've thought that to be a negative tradeoff on balance considering how security critical this is. More than happy to re-evaluate this down the line :)

Reply:
Mm yeah, I like this rationale - keeping in mind auditability as an axis along which to evaluate the code